### PR TITLE
Give the Schwarz subdomain operator the full power of the DataBox

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Just pass the full DataBox to the subdomain operator instead of calling it on the volume and on interfaces separately. This is needed so the subdomain operator can determine which boundary conditions were set by the user and retrieve their arguments dynamically (using the `call_with_dynamic_type` mechanism).

Since this removes the compile-time interface from the subdomain operator I switched the protocol to an abstract base class.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_SchwarzAlgorithm.cpp [hunk 1]
```diff
@@ -131,61 +118,50 @@ void restrict_to_subdomain(
...
-  template <typename Tag, typename ResultData, typename SubdomainOperator>
-  static void apply(
-      const std::vector<DenseMatrix<double>>& matrix_slices,
-      const Element<1>& element, const std::array<size_t, 1>& overlap_extents,
-      const LinearSolver::Schwarz::ElementCenteredSubdomainData<
-          1, tmpl::list<Tag>>& subdomain_data,
-      const gsl::not_null<ResultData*> result,
-      const gsl::not_null<SubdomainOperator*> /*subdomain_operator*/) noexcept {
+struct SubdomainOperator : LinearSolver::Schwarz::SubdomainOperator<1> {
+  // The operator can by non-const and can also mutate the DataBox to modify
```
Comment (moxcodes) [inline tests/Unit/ParallelAlgorithms/LinearSolver/Schwarz/Test_SchwarzAlgorithm.cpp]: `by`->`be`

File: src/ParallelAlgorithms/LinearSolver/Schwarz/SubdomainOperator.hpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ */
+template <size_t Dim>
+class SubdomainOperator {
+ public:
+  static constexpr size_t volume_dim = Dim;
+
+ protected:
+  SubdomainOperator() = default;
+
+  // Derived classes must implement operator() member function template
```
Comment (moxcodes) [inline src/ParallelAlgorithms/LinearSolver/Schwarz/SubdomainOperator.hpp]: I'm not sure whether this extra code comment adds much information given the detailed statements in the doxygen comment.

**Review by kidder — COMMENTED** [review]
I have no requested changes

**Comment by nilsvu** [comment]
Rebased and squashed in the two minor changes. Thanks for your reviews!

**Review by moxcodes — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**